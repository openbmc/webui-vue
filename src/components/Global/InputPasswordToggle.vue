<template>
  <div class="input-password-toggle-container">
    <slot></slot>
    <b-button
      :aria-label="togglePasswordLabel"
      :title="togglePasswordLabel"
      variant="link"
      class="input-action-btn btn-icon-only"
      :class="{ isVisible: isVisible }"
      @click="toggleVisibility"
    >
      <icon-view-off v-if="isVisible" />
      <icon-view v-else />
    </b-button>
  </div>
</template>

<script>
import IconView from '@carbon/icons-vue/es/view/20';
import IconViewOff from '@carbon/icons-vue/es/view--off/20';

export default {
  name: 'InputPasswordToggle',
  components: { IconView, IconViewOff },
  data() {
    return {
      isVisible: false,
      togglePasswordLabel: this.$t('global.ariaLabel.showPassword'),
    };
  },
  methods: {
    toggleVisibility() {
      const firstChild = this.$children[0];
      const inputEl = firstChild ? firstChild.$el : null;

      this.isVisible = !this.isVisible;

      if (inputEl && inputEl.nodeName === 'INPUT') {
        inputEl.type = this.isVisible ? 'text' : 'password';
      }

      this.isVisible
        ? (this.togglePasswordLabel = this.$t('global.ariaLabel.hidePassword'))
        : (this.togglePasswordLabel = this.$t('global.ariaLabel.showPassword'));
    },
  },
};
</script>

<style lang="scss" scoped>
.input-password-toggle-container {
  position: relative;
}
</style>
