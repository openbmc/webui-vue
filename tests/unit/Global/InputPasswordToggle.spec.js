import { mount, createLocalVue } from '@vue/test-utils';
import InputPasswordToggle from '@/components/Global/InputPasswordToggle';
import BootstrapVue from 'bootstrap-vue';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

describe('InputPasswordToggle.vue', () => {
  const wrapper = mount(InputPasswordToggle, {
    localVue,
    data() {
      return {
        isVisible: false,
      };
    },
    mocks: {
      $t: (key) => key,
    },
  });
  it('should exist', () => {
    expect(wrapper.exists()).toBe(true);
  });
  it('should not render isVisible class', () => {
    expect(wrapper.find('.isVisible').exists()).toBe(false);
  });
  it('should render isVisible class when button is clicked', async () => {
    await wrapper.find('button').trigger('click');
    expect(wrapper.find('.isVisible').exists()).toBe(true);
  });
  it('should render correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
