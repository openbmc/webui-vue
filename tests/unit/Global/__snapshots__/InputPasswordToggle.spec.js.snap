// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputPasswordToggle.vue should render correctly 1`] = `
<div
  class="input-password-toggle-container"
>
  
  
  <b-button
    class="input-action-btn btn-icon-only"
    title="Show password as plain text. Note: this will visually expose your password on the screen."
    variant="link"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      focusable="false"
      height="20"
      preserveAspectRatio="xMidYMid meet"
      viewBox="0 0 32 32"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M30.94,15.66A16.69,16.69,0,0,0,16,5,16.69,16.69,0,0,0,1.06,15.66a1,1,0,0,0,0,.68A16.69,16.69,0,0,0,16,27,16.69,16.69,0,0,0,30.94,16.34,1,1,0,0,0,30.94,15.66ZM16,25c-5.3,0-10.9-3.93-12.93-9C5.1,10.93,10.7,7,16,7s10.9,3.93,12.93,9C26.9,21.07,21.3,25,16,25Z"
      />
      <path
        d="M16,10a6,6,0,1,0,6,6A6,6,0,0,0,16,10Zm0,10a4,4,0,1,1,4-4A4,4,0,0,1,16,20Z"
      />
    </svg>
    <span
      class="sr-only"
    >
      Show password as plain text. Note: this will visually expose your password on the screen.
    </span>
  </b-button>
</div>
`;
